File: servidor.js
/* Servidor local mínimo para el Editor PDF Clarvi (alternativa a servidor.py).
   No instala nada ni sale a internet: sólo sirve esta carpeta en 127.0.0.1. */
'use strict';

const http = require('http');
const fs = require('fs');
const path = require('path');
const { exec } = require('child_process');

const RAIZ = __dirname;
const PUERTO_INICIAL = 8756;

const TIPOS = {
  '.html': 'text/html; charset=utf-8',
  '.css': 'text/css; charset=utf-8',
  '.js': 'text/javascript; charset=utf-8',
  '.json': 'application/json; charset=utf-8',
  '.pdf': 'application/pdf',
  '.png': 'image/png',
  '.jpg': 'image/jpeg',
  '.svg': 'image/svg+xml',
  '.bcmap': 'application/octet-stream',
  '.pfb': 'application/octet-stream',
  '.ttf': 'font/ttf'
};

const servidor = http.createServer((req, res) => {
  let ruta = decodeURIComponent(req.url.split('?')[0]);
  if (ruta === '/') ruta = '/index.html';

  const destino = path.join(RAIZ, path.normalize(ruta));
  if (!destino.startsWith(RAIZ)) { res.writeHead(403).end('Prohibido'); return; }

  fs.readFile(destino, (err, datos) => {
    if (err) { res.writeHead(404).end('No encontrado'); return; }
    res.writeHead(200, {
      'Content-Type': TIPOS[path.extname(destino).toLowerCase()] || 'application/octet-stream',
      'Cache-Control': 'no-store'
    });
    res.end(datos);
  });
});

function abrirNavegador(url) {
  const orden = process.platform === 'win32' ? `start "" "${url}"`
              : process.platform === 'darwin' ? `open "${url}"`
              : `xdg-open "${url}"`;
  exec(orden, () => {});
}

function escuchar(puerto, intentos) {
  servidor.once('error', (e) => {
    if (e.code === 'EADDRINUSE' && intentos > 0) escuchar(puerto + 1, intentos - 1);
    else { console.error('No se pudo iniciar el servidor:', e.message); process.exit(1); }
  });
  servidor.listen(puerto, '127.0.0.1', () => {
    const url = `http://127.0.0.1:${puerto}/index.html`;
    console.log('');
    console.log('  Editor PDF Clarvi');
    console.log('  -----------------');
    console.log('  Abierto en: ' + url);
    console.log('  Deja esta ventana abierta mientras lo uses.');
    console.log('  Para cerrarlo: Ctrl+C');
    console.log('');
    setTimeout(() => abrirNavegador(url), 500);
  });
}

escuchar(PUERTO_INICIAL, 30);

File: servidor.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Servidor local mínimo para el Editor PDF Clarvi.

No instala nada ni sale a internet: sólo sirve los archivos de esta carpeta
en 127.0.0.1 para que el navegador pueda usar el worker de pdf.js y las
fuentes estándar. Ciérralo con Ctrl+C cuando termines.
"""
import http.server
import os
import socketserver
import sys
import threading
import webbrowser

PUERTO_INICIAL = 8756
RAIZ = os.path.dirname(os.path.abspath(__file__))


class Manejador(http.server.SimpleHTTPRequestHandler):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, directory=RAIZ, **kwargs)

    def log_message(self, formato, *args):
        pass  # sin ruido en la consola

    def end_headers(self):
        self.send_header('Cache-Control', 'no-store')
        super().end_headers()


def main():
    puerto = PUERTO_INICIAL
    servidor = None
    for _ in range(30):
        try:
            socketserver.TCPServer.allow_reuse_address = True
            servidor = socketserver.TCPServer(('127.0.0.1', puerto), Manejador)
            break
        except OSError:
            puerto += 1

    if servidor is None:
        print('No se encontró ningún puerto libre. Abre index.html directamente.')
        return 1

    url = 'http://127.0.0.1:%d/index.html' % puerto
    print('')
    print('  Editor PDF Clarvi')
    print('  -----------------')
    print('  Abierto en: %s' % url)
    print('  Deja esta ventana abierta mientras lo uses.')
    print('  Para cerrarlo: Ctrl+C')
    print('')

    threading.Timer(0.7, lambda: webbrowser.open(url)).start()
    try:
        servidor.serve_forever()
    except KeyboardInterrupt:
        print('\n  Servidor detenido. ¡Hasta luego!')
    finally:
        servidor.server_close()
    return 0


if __name__ == '__main__':
    sys.exit(main())
